format-delimited/src/main/java/io/cdap/plugin/format/delimited/input/PathTrackingDelimitedInputFormat.java

```diff
@@ -30,16 +31,82 @@
 import org.apache.hadoop.mapreduce.lib.input.FileSplit;
 
 import java.io.IOException;
+import java.util.ArrayList;
 import java.util.Iterator;
+import java.util.List;
 import javax.annotation.Nullable;
 
 /**
  * Delimited text format that tracks which file each record was read from.
  */
 public class PathTrackingDelimitedInputFormat extends PathTrackingInputFormat {
   static final String DELIMITER = "delimiter";
+  static final String ENABLE_QUOTES_VALUE = "enable_quotes_value";
   static final String SKIP_HEADER = "skip_header";
 
+  private static final String QUOTE = "\"";
+  private static final char QUOTE_CHAR = '\"';
+
+  /**
+   * Split the delimited string based on the delimiter. The delimiter should not contain any quotes.
+   * The method will behave like this: 1. if there is no quote, it will behave same as {@link
+   * String#split(String)} 2. if there are quotes in the string, the method will find pairs of
+   * quotes, content within each pair of quotes will not get splitted even if there is delimiter in
+   * that. For example, if string is a."b.c"."d.e.f" and delimiter is '.', it will get split into
+   * [a, b.c, d.e.f]. if string is "val1.val2", then it will not get splitted since the '.' is
+   * within pair of quotes. If the delimited string contains odd number of quotes, which mean the
+   * quotes are not closed, an exception will be thrown. The quote within the value will always be
+   * trimed.
+   *
+   * @param delimitedString the string to split
+   * @param delimiter the separtor
+   * @return a list of splits of the original string
+   */
+  @VisibleForTesting
+  static Iterable<String> splitQuotesString(String delimitedString, String delimiter)
+    throws IOException {
+
+    boolean isWithinQuotes = false;
+    List<String> result = new ArrayList<>();
```

Comment:
since the results are stored in a List, this actually ends up looping twice for each line. Once to build this list and then once again to go through the List iterator.

It is better to create something like a SplitQuotesIterator that has mostly the same logic, except it keeps track of i and only returns the next value when needed. This way it only loops once. You can extend AbstractIterator from guava to simplify the implementation.